src/basic/fileio.c

```diff
@@ -755,9 +755,12 @@ int read_line_full(FILE *f, size_t limit, ReadLineFlags flags, char **ret) {
                             (eol == EOL_NONE && previous_eol != EOL_NONE) ||
                             (eol != EOL_NONE && (previous_eol & eol) != 0)) {
                                 /* Previous char was a NUL? This is not an EOL, but the previous char was? This type of
-                                 * EOL marker has been seen right before? In either of these three cases we are
-                                 * done. But first, let's put this character back in the queue. */
-                                assert_se(ungetc(c, f) != EOF);
+                                 * EOL marker has been seen right before?  In either of these three cases we are
+                                 * done. But first, let's put this character back in the queue. (Note that we have to
+                                 * cast this to (unsigned char) here as fungetc() expects a positive 'int', and if we
+                                 * are on an architecture where 'char' equals 'signed char' we need to ensure we don't
+                                 * pass a negative value here.) */
+                                assert_se(ungetc((unsigned char) c, f) != EOF);
```

Comment:
Hmm, I thought I added a test exactly for this. Adding the cast is OK, but I'd like to understand why the test in `test-fileio` didn't catch this.

Reply:
Ah, OK, so `ungetc` fails with `\377`. Why do all libc interfaces have to be such shit?

Reply:
i figure it's because you didn't add 0xff to the list of chars...

ungetc() doesn#t really care too much about whatever it gets, except if it gets EOF which is -1... hence, passing in -1 is never ok...

Reply:
yes, i verified... if \xff (aka \377) is added to the safe_fgetc() test, then it fails)

Reply:
Hmm, so maybe adjust the comment?